packages/collector/src/index.ts

```diff
@@ -1 +1,211 @@
+      status = PositionStatus.New
+      break
+
+    case EventType.Unwind:
+      // event.args[0] = sender
+      // event.args[1] = positionId
+      // event.args[2] = fraction
+      positionId = ethers.BigNumber.from(event.args[1]).toString()
+      // if the position is fully unwound, remove it from the cache
+      if (event.args[2] === '1000000000000000000') {
+        await redis.hdel(`positions:${marketAddress}`, positionId)
+        await redis.zrem(`position_index:${marketAddress}`, positionId);
+        status = PositionStatus.Removed
+      } else {
+        status = PositionStatus.Updated;
+      }
+      break
+
+    case EventType.Liquidate:
+      // event.args[0] = sender
+      // event.args[1] = owner
+      // event.args[2] = positionId
+      positionId = ethers.BigNumber.from(event.args[2]).toString()
+      await redis.hdel(`positions:${marketAddress}`, positionId)
+      await redis.zrem(`position_index:${marketAddress}`, positionId);
+      status = PositionStatus.Removed
+      break
+
+    default:
+      log(chalk.bold.red(`Unhandled event type: ${eventName}`))
+      status = PositionStatus.Error
+      break
+  }
+
+  await redis.set(`LatestBlockProcessed:${marketAddress}`, (event.blockNumber + 1).toString())
+  return status
+}
+
+// Fetch events for a given market
+async function fetchEvents(marketName: string) {
+  const marketAddress = markets[marketName]
+  const provider = new ethers.providers.JsonRpcProvider(process.env.RPC_URLS?.split(',')[0])
+  const ovlMarketContract = new ethers.Contract(marketAddress, market_abi, provider)
+
+  // get the latest block processed for the market
+  let startBlock = await redis.get(`LatestBlockProcessed:${marketAddress}`)
+  // get the latest block from the RPC provider
+  const latestBlock = await provider.getBlockNumber()
+  // current block step to fetch events
+  const blockStep = 45000 // adjust as per your RPC limitations
+
+  let allEvents: ethers.Event[] = []
+
+  // if the start block is not found in Redis, get events from the block where the market was deployed to the latest block
+  // or if the difference between the latest block and the start block is greater than the block step
+  if (!startBlock || latestBlock - parseInt(startBlock) > blockStep) {
+    startBlock = startBlock || markets_block[marketName]
+
+    log(
+      `Getting events from block: ${chalk.green(startBlock)} to block: ${chalk.green(latestBlock)} for market: ${chalk.bold.blue(marketName)}`
+    )
+
+    // create an array to hold promises for each block range
+    const promises = []
+    for (let block = parseInt(startBlock); block < latestBlock; block += blockStep + 1) {
+      const fromBlock = block
+      const toBlock = Math.min(block + blockStep, latestBlock)
+      promises.push(ovlMarketContract.queryFilter('*', fromBlock, toBlock))
+    }
+
+    // execute all promises in parallel
+    const eventsArrays = await Promise.all(promises)
```

Comment:
Could be interesting to manage partial success using [Promise.allSettled()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) instead of Promise.all.

Reply:
In the other hand, implementing allSettled here is not so good because here is fetching the events by batch of 45000 blocks, so if one of the batches fails, I would like to not process the others.